e2e/cluster.go

```diff
@@ -2,14 +2,256 @@ package e2e
+	maxRetries := 10
+	retryInterval := 30 * time.Second
+	var lastErr error
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		log.Printf("Attempt %d: creating or updating cluster %s in rg %s\n", attempt+1, *cluster.Name, *cluster.Location)
+
+		createdCluster, err := createNewAKSCluster(ctx, cluster)
+		if err == nil {
+			return createdCluster, nil
+		}
+
+		// Check if the error is a 409 Conflict
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 409 {
+			lastErr = err
+			log.Printf("Attempt %d failed with 409 Conflict: %v. Retrying in %v...\n", attempt+1, err, retryInterval)
+
+			select {
+			case <-time.After(retryInterval):
+				// Continue to next iteration
+			case <-ctx.Done():
+				return nil, fmt.Errorf("context canceled while retrying cluster creation: %w", ctx.Err())
+			}
+		} else {
+			// If it's not a 409 error, return immediately
+			return nil, fmt.Errorf("failed to create cluster: %w", err)
+		}
+	}
+
+	return nil, fmt.Errorf("failed to create cluster after %d attempts due to persistent 409 Conflict: %w", maxRetries, lastErr)
+}
+
+type VNet struct {
+	name     string
+	subnetId string
+}
+
+func getClusterVNet(ctx context.Context, mcResourceGroupName string) (VNet, error) {
+	pager := config.Azure.VNet.NewListPager(mcResourceGroupName, nil)
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			return VNet{}, fmt.Errorf("failed to advance page: %w", err)
+		}
+		for _, v := range nextResult.Value {
+			if v == nil {
+				return VNet{}, fmt.Errorf("aks vnet was empty")
+			}
+			return VNet{name: *v.Name, subnetId: fmt.Sprintf("%s/subnets/%s", *v.ID, "aks-subnet")}, nil
+		}
+	}
+	return VNet{}, fmt.Errorf("failed to find aks vnet")
+}
+
+func collectGarbageVMSS(ctx context.Context, cluster *armcontainerservice.ManagedCluster) error {
+	rg := *cluster.Properties.NodeResourceGroup
+	pager := config.Azure.VMSS.NewListPager(rg, nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get next page of VMSS: %w", err)
+		}
+		for _, vmss := range page.Value {
+			if _, ok := vmss.Tags["KEEP_VMSS"]; ok {
+				continue
+			}
+			// don't delete managed pools
+			if _, ok := vmss.Tags["aks-managed-poolName"]; ok {
+				continue
+			}
+
+			// don't delete VMSS created in the last hour. They might be currently used in tests
+			// extra 10 minutes is to avoid clock drift issues between the test machine and the Azure API
+			if config.Timeout == 0 || time.Since(*vmss.Properties.TimeCreated) < config.Timeout+10*time.Minute {
+				continue
+			}
+
+			_, err := config.Azure.VMSS.BeginDelete(ctx, rg, *vmss.Name, &armcompute.VirtualMachineScaleSetsClientBeginDeleteOptions{
+				ForceDeletion: to.Ptr(true),
+			})
+			if err != nil {
+				log.Printf("failed to delete vmss %q: %s", *vmss.Name, err)
+			}
+			log.Printf("deleted garbage vmss %q", *vmss.ID)
```

Comment:
should this log line be in an else clause to avoid logging false information ?

Reply:
oopsie, a classic case of wrong error handling